File: Cpp/AdaBoost.cpp
#include <iostream>
#include <cmath>
#include <vector>
#include <limits>
#include <exception>

#include "Stump.h"
#include "Data.h"

class AdaBoost {
	double* alphas = nullptr;
	std::vector<stump> h;

	unsigned predict(double* obs) {
		double max_val = std::numeric_limits<double>::min();
		unsigned prediction = CLASS_NR;
		for (unsigned p = 0; p < CLASS_NR; p++) {
			double val = 0.0;
			for (unsigned i = 0; i < h.size(); i++) 
				val += (h[i].predict(obs) == p) ? alphas[i] : 0.0;
			if (val > max_val) {
				max_val = val;
				prediction = p;
			}
		}
		return prediction;
	}
public:
	~AdaBoost() {
		delete[]alphas;
	}

	void fit(double** x, unsigned* y, unsigned iter_nr = 1000, unsigned size = TRAIN_SIZE) {
		h.reserve(iter_nr);
		double* weights = new double[size];
		alphas = new double[iter_nr];
		double n = static_cast<double>(size);
		for (unsigned i = 0; i < size; i++)
			weights[i] = 1.0 / n;
		for (unsigned t = 0; t < iter_nr; t++) {
			stump st;
			double err = st.fit(x, y, weights, size);
			double alpha = log((1.0 - err) / err) + log(static_cast<double>(CLASS_NR) - 1.0);
			alphas[t] = alpha;
			for (unsigned i = 0; i < size; i++)
				weights[i] *= exp(alpha * (st.predict(x[i]) != y[i]) ? 1.0 : -1.0);
			double sum = 0.0;
			for (unsigned i = 0; i < size; i++)
				sum += weights[i];
			for (unsigned i = 0; i < size; i++)
				weights[i] /= sum;
			h.push_back(st);
		}
		delete[]weights;
	}

	double check(double** x, unsigned* y, unsigned size = TEST_SIZE) {
		if (h.empty())
			throw std::runtime_error("Error: check called before fitting the model.");
		int correct = 0;
		for (unsigned i = 0; i < size; i++)
			if (predict(x[i]) == y[i])
				correct++;
		return static_cast<double>(correct) * 100.0 / static_cast<double>(size);
	}
};

int main() {
	set_holder data = prepare_data();
	AdaBoost ad;
	try {
		ad.fit(data.train_x, data.train_y);
		std::cout << ad.check(data.test_x, data.test_y) << std::endl;
	}
	catch(const std::runtime_error& er) {
		std::cerr << er.what() << std::endl;
	}
	data.clear();
}

File: Python/AdaBoost.py
import numpy as np

from DataPreparation import SetHolder
from sklearn.tree import DecisionTreeClassifier


class AdaBoost:
    def __init__(self):
        self.alphas = None
        self.h = []
        self.class_nr = 0

    def fit(self, x, y, iter_nr=50):
        self.h = []
        self.alphas = np.empty(shape=(iter_nr, ))
        self.class_nr = len(np.unique(y))
        weights = np.full((len(x), ), 1 / len(x))
        for t in range(iter_nr):
            st = DecisionTreeClassifier(max_depth=1)
            st.fit(x, y, sample_weight=weights)
            y_pred = st.predict(x)
            err = np.sum(weights * np.not_equal(y, y_pred).astype(int))
            alpha = np.log((1.0 - err) / err) + np.log(self.class_nr - 1)
            self.alphas[t] = alpha
            weights *= np.exp(alpha * np.not_equal(y, y_pred).astype(int))
            weights /= np.sum(weights)
            self.h.append(st)

    def _predict(self, obs):
        prediction = 0
        max_val = 0.0
        for p in range(self.class_nr):
            val = 0.0
            for stump, alpha in zip(self.h, self.alphas):
                if stump.predict(obs.reshape(1, -1))[0] == p:
                    val += alpha
            if val > max_val:
                max_val = val
                prediction = p
        return prediction

    def check(self, x, y):
        y_pred = np.empty(shape=(len(x), ), dtype=int)
        for i, val in enumerate(x):
            y_pred[i] = self._predict(val)
        return np.sum(np.equal(y, y_pred).astype(int)) / len(y) * 100.0


if __name__ == '__main__':
    data = SetHolder()
    ad = AdaBoost()
    ad.fit(data.train_x, data.train_y, 100)
    ac = ad.check(data.test_x, data.test_y)
    print(ac)
